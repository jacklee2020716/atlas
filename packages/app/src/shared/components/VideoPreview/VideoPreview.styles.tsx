import styled from '@emotion/styled'
import { colors, spacing, typography } from '../../theme'
import Avatar from '../Avatar'
import { PlayIcon } from '../../icons'

const HOVER_BORDER_SIZE = '2px'

type CoverImageProps = {
  displayPosterPlaceholder?: boolean
}

type ContainerProps = {
  clickable: boolean
}

type ChannelProps = {
  channelClickable: boolean
}

export const CoverContainer = styled.div`
  width: 320px;
  height: 190px;

  transition-property: box-shadow, transform;
  transition-duration: 0.4s;
  transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1);

  position: relative;
`

export const CoverImage = styled.img<CoverImageProps>`
  width: 100%;
  height: 100%;
  background-image: ${({ displayPosterPlaceholder }) =>
    displayPosterPlaceholder ? `linear-gradient(${colors.gray[300]}, ${colors.gray[700]})` : 'none'};
  background-size: cover;
  object-fit: cover;
`

export const CoverHoverOverlay = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  opacity: 0;

  transition: opacity 0.4s cubic-bezier(0.165, 0.84, 0.44, 1);

  border: ${HOVER_BORDER_SIZE} solid ${colors.white};
  background: linear-gradient(180deg, #000000 0%, rgba(0, 0, 0, 0) 100%);

  display: flex;
  justify-content: center;
  align-items: center;
`

export const CoverPlayIcon = styled(PlayIcon)`
  transition: transform 0.4s cubic-bezier(0.165, 0.84, 0.44, 1);
  transform: translateY(40px);
  width: 54px;
  height: 54px;
`

export const ProgressOverlay = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  height: ${spacing.xxs};
  background-color: ${colors.white};
`

export const ProgressBar = styled.div`
  position: absolute;
  left: 0;
  bottom: 0;
  height: 100%;
  width: 0;
  background-color: ${colors.blue['500']};
`

export const Container = styled.div<ContainerProps>`
  color: ${colors.gray[300]};
  cursor: ${({ clickable }) => (clickable ? 'pointer' : 'auto')};
  display: inline-block;
  ${({ clickable }) =>
    clickable &&
    `
				&:hover {
					${CoverContainer} {
						transform: translate(-${spacing.xs}, -${spacing.xs});
						box-shadow: ${spacing.xs} ${spacing.xs} 0 ${colors.blue['500']};
					}

					${CoverHoverOverlay} {
						opacity: 1;
					}
					
					${CoverPlayIcon} {
						transform: translateY(0);
					}

					${ProgressOverlay} {
						bottom: ${HOVER_BORDER_SIZE};
					}
				}
			`}
`

export const CoverDurationOverlay = styled.div`
  position: absolute;
  bottom: ${spacing.xs};
  right: ${spacing.xs};
  padding: ${spacing.xxs} ${spacing.xs};
  background-color: ${colors.black};
  color: ${colors.white};
  font-size: ${typography.sizes.body2};
`

export const InfoContainer = styled.div`
  display: flex;
  margin-top: ${spacing.s};
`

export const StyledAvatar = styled(Avatar)<ChannelProps>`
  width: 40px;
  height: 40px;
  margin-right: ${spacing.xs};
  cursor: ${({ channelClickable }) => (channelClickable ? 'pointer' : 'auto')};
`

export const TextContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: start;
`

export const TitleHeader = styled.h3`
  margin: 0;
  font-weight: ${typography.weights.bold};
  font-size: ${typography.sizes.h6};
  line-height: 1.25rem;
  color: ${colors.white};
  display: inline-block;
`

export const ChannelName = styled.span<ChannelProps>`
  font-size: ${typography.sizes.subtitle2};
  line-height: 1.25rem;
  display: inline-block;
  cursor: ${({ channelClickable }) => (channelClickable ? 'pointer' : 'auto')};
`

export const MetaText = styled.span`
  margin-top: ${spacing.xs};
  font-size: ${typography.sizes.subtitle2};
`
