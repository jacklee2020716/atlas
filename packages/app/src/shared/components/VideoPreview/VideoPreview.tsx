import React from 'react'
import {
  ChannelName,
  CoverDurationOverlay,
  CoverHoverOverlay,
  CoverImage,
  CoverPlayIcon,
  MetaText,
  ProgressBar,
  ProgressOverlay,
  StyledAvatar,
  TitleHeader,
} from './VideoPreview.styles'
import { formatDateAgo, formatDurationShort } from '@/utils/time'
import { formatNumberShort } from '@/utils/number'
import VideoPreviewBase from './VideoPreviewBase'

type VideoPreviewProps = {
  title: string
  channelName: string
  channelAvatarURL?: string | null
  createdAt: Date
  duration?: number
  // video watch progress in percent (0-100)
  progress?: number
  views: number
  posterURL: string

  showChannel?: boolean
  showMeta?: boolean
  imgRef?: React.Ref<HTMLImageElement>
  onClick?: (e: React.MouseEvent<HTMLElement>) => void
  onChannelClick?: (e: React.MouseEvent<HTMLElement>) => void
  className?: string
}

const VideoPreview: React.FC<VideoPreviewProps> = ({
  title,
  channelName,
  channelAvatarURL,
  createdAt,
  duration,
  progress = 0,
  views,
  posterURL,
  showChannel = true,
  showMeta = true,
  imgRef,
  onClick,
  onChannelClick,
  className,
}) => {
  const channelClickable = !!onChannelClick

  const handleChannelClick = (e: React.MouseEvent<HTMLElement>) => {
    if (!onChannelClick) {
      return
    }
    e.stopPropagation()
    onChannelClick(e)
  }

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (!onClick) {
      return
    }
    e.stopPropagation()
    onClick(e)
  }

  const coverNode = (
    <>
      <CoverImage src={posterURL} ref={imgRef} alt={`${title} by ${channelName} thumbnail`} />
      {duration && <CoverDurationOverlay>{formatDurationShort(duration)}</CoverDurationOverlay>}
      {!!progress && (
        <ProgressOverlay>
          <ProgressBar style={{ width: `${progress}%` }} />
        </ProgressOverlay>
      )}
      <CoverHoverOverlay>
        <CoverPlayIcon />
      </CoverHoverOverlay>
    </>
  )

  const titleNode = <TitleHeader>{title}</TitleHeader>

  const channelAvatarNode = (
    <StyledAvatar
      size="small"
      name={channelName}
      img={channelAvatarURL}
      channelClickable={channelClickable}
      onClick={handleChannelClick}
    />
  )

  const channelNameNode = (
    <ChannelName channelClickable={channelClickable} onClick={handleChannelClick}>
      {channelName}
    </ChannelName>
  )

  const metaNode = (
    <MetaText>
      {formatDateAgo(createdAt)}ãƒ»{formatNumberShort(views)} views
    </MetaText>
  )

  return (
    <VideoPreviewBase
      coverNode={coverNode}
      titleNode={titleNode}
      showChannel={showChannel}
      channelAvatarNode={channelAvatarNode}
      channelNameNode={channelNameNode}
      showMeta={showMeta}
      metaNode={metaNode}
      onClick={handleClick}
      className={className}
    />
  )
}

export default VideoPreview
