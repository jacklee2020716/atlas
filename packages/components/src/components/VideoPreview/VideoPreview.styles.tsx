import { css } from "@emotion/core";
import { typography, colors } from "./../../theme";

export type VideoPreviewStyleProps = {
	showChannel?: boolean;
};

export let makeStyles = ({ showChannel = false }: VideoPreviewStyleProps) => {
	return {
		container: css``,
		link: css`
			text-decoration: none;
		`,
		coverContainer: css`
			width: 100%;
			background-color: black;
		`,
		cover: css`
			display: block;
			width: 100%;
			height: auto;
		`,
		infoContainer: css`
			display: grid;
			grid-template: auto / ${showChannel ? "45px auto" : "auto"};
			margin: 10px 0 0;
		`,
		avatar: css`
			grid-column: 1 / 1;
		`,
		textContainer: css`
			grid-column: ${showChannel ? "2 / 2" : "1 / 1"};
		`,
		title: css`
			margin: 0;
			font-weight: ${typography.weights.bold};
			text-transform: capitalize;
			color: ${colors.black};
			font-size: ${typography.sizes.h3};
		`,
		channel: css`
			margin: 5px 0 0;
			font-size: ${typography.sizes.subtitle2};
			color: ${colors.gray[700]};
		`,
	};
};
