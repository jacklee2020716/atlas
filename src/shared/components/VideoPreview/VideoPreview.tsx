import React, { useState } from 'react'
import useResizeObserver from 'use-resize-observer'

import {
  ChannelName,
  CoverDurationOverlay,
  CoverHoverOverlay,
  CoverImage,
  CoverPlayIcon,
  MetaText,
  ProgressBar,
  ProgressOverlay,
  StyledAvatar,
  TitleHeader,
} from './VideoPreview.styles'
import { formatDurationShort } from '@/utils/time'
import VideoPreviewBase from './VideoPreviewBase'
import { formatVideoViewsAndDate } from '@/utils/video'

const MIN_COVER_WIDTH = 300
const MAX_COVER_WIDTH = 600
const MIN_SCALING_FACTOR = 1
const MAX_SCALING_FACTOR = 1.375
// Linear Interpolation, see https://en.wikipedia.org/wiki/Linear_interpolation
const calculateScalingFactor = (coverWidth: number) =>
  MIN_SCALING_FACTOR +
  ((coverWidth - MIN_COVER_WIDTH) * (MAX_SCALING_FACTOR - MIN_SCALING_FACTOR)) / (MAX_COVER_WIDTH - MIN_COVER_WIDTH)

type VideoPreviewProps = {
  title: string
  channelName: string
  channelAvatarURL?: string | null
  createdAt: Date
  duration?: number
  // video watch progress in percent (0-100)
  progress?: number
  views: number | null
  posterURL: string

  showChannel?: boolean
  showMeta?: boolean
  main?: boolean

  imgRef?: React.Ref<HTMLImageElement>
  onClick?: (e: React.MouseEvent<HTMLElement>) => void
  onChannelClick?: (e: React.MouseEvent<HTMLElement>) => void
  className?: string
}

const VideoPreview: React.FC<VideoPreviewProps> = ({
  title,
  channelName,
  channelAvatarURL,
  createdAt,
  duration,
  progress = 0,
  views,
  posterURL,
  showChannel = true,
  showMeta = true,
  main = false,
  onClick,
  onChannelClick,
  className,
}) => {
  const [scalingFactor, setScalingFactor] = useState(MIN_SCALING_FACTOR)
  const { ref: imgRef } = useResizeObserver<HTMLImageElement>({
    onResize: (size) => {
      const { width: coverWidth } = size
      if (coverWidth && !main) {
        setScalingFactor(calculateScalingFactor(coverWidth))
      }
    },
  })

  const channelClickable = !!onChannelClick

  const handleChannelClick = (e: React.MouseEvent<HTMLElement>) => {
    if (!onChannelClick) {
      return
    }
    e.stopPropagation()
    onChannelClick(e)
  }

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (!onClick) {
      return
    }
    e.stopPropagation()
    onClick(e)
  }

  const coverNode = (
    <>
      <CoverImage src={posterURL} ref={imgRef} alt={`${title} by ${channelName} thumbnail`} />
      {duration && <CoverDurationOverlay>{formatDurationShort(duration)}</CoverDurationOverlay>}
      {!!progress && (
        <ProgressOverlay>
          <ProgressBar style={{ width: `${progress}%` }} />
        </ProgressOverlay>
      )}
      <CoverHoverOverlay>
        <CoverPlayIcon />
      </CoverHoverOverlay>
    </>
  )

  const titleNode = (
    <TitleHeader main={main} scalingFactor={scalingFactor}>
      {title}
    </TitleHeader>
  )

  const channelAvatarNode = (
    <StyledAvatar
      size="small"
      name={channelName}
      img={channelAvatarURL}
      channelClickable={channelClickable}
      onClick={handleChannelClick}
    />
  )

  const channelNameNode = (
    <ChannelName channelClickable={channelClickable} onClick={handleChannelClick} scalingFactor={scalingFactor}>
      {channelName}
    </ChannelName>
  )

  const metaNode = (
    <MetaText main={main} scalingFactor={scalingFactor}>
      {formatVideoViewsAndDate(views, createdAt, { fullViews: main })}
    </MetaText>
  )

  return (
    <VideoPreviewBase
      coverNode={coverNode}
      titleNode={titleNode}
      showChannel={showChannel}
      channelAvatarNode={channelAvatarNode}
      channelNameNode={channelNameNode}
      showMeta={showMeta}
      main={main}
      metaNode={metaNode}
      onClick={onClick && handleClick}
      className={className}
      scalingFactor={scalingFactor}
    />
  )
}

export default VideoPreview
