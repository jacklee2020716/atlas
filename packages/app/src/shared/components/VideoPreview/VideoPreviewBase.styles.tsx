import styled from '@emotion/styled'
import { colors, spacing } from '@/shared/theme'
import { CoverHoverOverlay, CoverPlayIcon, ProgressOverlay } from './VideoPreview.styles'

export const HOVER_BORDER_SIZE = '2px'

type ContainerProps = {
  clickable: boolean
}

export const MAX_VIDEO_PREVIEW_WIDTH = '320px'

export const CoverContainer = styled.div`
  width: 100%;
  height: 190px;

  transition-property: box-shadow, transform;
  transition-duration: 0.4s;
  transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1);

  position: relative;
`

export const Container = styled.article<ContainerProps>`
  max-width: ${MAX_VIDEO_PREVIEW_WIDTH};
  color: ${colors.gray[300]};
  cursor: ${({ clickable }) => (clickable ? 'pointer' : 'auto')};
  display: inline-block;
  ${({ clickable }) =>
    clickable &&
    `
				&:hover {
					${CoverContainer} {
						transform: translate(-${spacing.xs}, -${spacing.xs});
						box-shadow: ${spacing.xs} ${spacing.xs} 0 ${colors.blue['500']};
					}

					${CoverHoverOverlay} {
						opacity: 1;
					}
					
					${CoverPlayIcon} {
						transform: translateY(0);
					}

					${ProgressOverlay} {
						bottom: ${HOVER_BORDER_SIZE};
					}
				}
			`}
`

export const InfoContainer = styled.div`
  display: flex;
  margin-top: ${spacing.s};
`

export const AvatarContainer = styled.div`
  width: 40px;
  min-width: 40px;
  height: 40px;
  margin-right: ${spacing.xs};
`

export const TextContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: start;
  width: 100%;
`

export const MetaContainer = styled.div`
  margin-top: ${spacing.xs};
  width: 100%;
`
