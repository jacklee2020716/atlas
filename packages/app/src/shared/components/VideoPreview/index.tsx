import VideoPreview from './VideoPreview'
import VideoPreviewBase from './VideoPreviewBase'
import { MAX_VIDEO_PREVIEW_WIDTH } from './VideoPreviewBase.styles'

export { VideoPreview, VideoPreviewBase, MAX_VIDEO_PREVIEW_WIDTH }
