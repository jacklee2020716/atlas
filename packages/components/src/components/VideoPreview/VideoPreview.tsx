import React from "react"

import { makeStyles, VideoPreviewStyleProps } from "./VideoPreview.styles"
import Avatar from "./../Avatar"

type VideoPreviewProps = {
  title: string
  channel?: string
  channelImg?: string
  showChannel?: boolean
  poster?: string
  onClick?: any
  onChannelClick?: any
} & VideoPreviewStyleProps

export default function VideoPreview({
  title,
  channel,
  channelImg,
  showChannel = false,
  poster,
  onClick,
  onChannelClick,
  ...styleProps
}: VideoPreviewProps) {
  let styles = makeStyles({ showChannel, ...styleProps })
  return (
    <div css={styles.container} onClick={onClick}>
      <div css={styles.coverContainer}>
        <img css={styles.cover} src={poster} onClick={event => { event.stopPropagation(); onClick() }} />
      </div>
      <div css={styles.infoContainer}>
        {showChannel && (
          <div css={styles.avatar} onClick={event => { event.stopPropagation(); onChannelClick() }}>
            <Avatar size="small" img={channelImg} />
          </div>
        )}
        <div css={styles.textContainer}>
          <h3 css={styles.title} onClick={event => { event.stopPropagation(); onClick() }}>{title}</h3>
          {showChannel && (
            <h3 css={styles.channel} onClick={event => { event.stopPropagation(); onChannelClick() }}>{channel}</h3>
          )}
        </div>
      </div>
    </div>
  )
}
