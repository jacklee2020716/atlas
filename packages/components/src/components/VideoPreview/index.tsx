import { memo } from "react"
import VideoPreview from "./VideoPreview"
export default memo(VideoPreview)
