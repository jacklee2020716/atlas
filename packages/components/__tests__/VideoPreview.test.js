import React from "react"
import { mount } from "enzyme"
import VideoPreview from "./../src/components/VideoPreview"

describe("VideoPreview component", () => {

  const component = mount(
    <VideoPreview
      title="Test"
      poster="https://ssl-static.libsyn.com/p/assets/a/4/8/f/a48f1a0697e958ce/Cover_2.png"
    />)

  it("Should render correctly", () => {
    expect(component).toMatchSnapshot()
  })

})