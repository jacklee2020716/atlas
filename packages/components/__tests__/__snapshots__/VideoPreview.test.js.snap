// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPreview component Should render correctly 1`] = `
<Memo(VideoPreview)
  poster="https://ssl-static.libsyn.com/p/assets/a/4/8/f/a48f1a0697e958ce/Cover_2.png"
  title="Test"
>
  <div
    css={
      Object {
        "map": undefined,
        "name": "0",
        "next": undefined,
        "styles": "",
        "toString": [Function],
      }
    }
  >
    <div
      css={
        Object {
          "map": undefined,
          "name": "fqj6m7",
          "next": undefined,
          "styles": "
      width: 100%;
      background-color: black;
    ",
          "toString": [Function],
        }
      }
    >
      <img
        css={
          Object {
            "map": undefined,
            "name": "1lk1tws",
            "next": undefined,
            "styles": "
      display: block;
      width: 100%;
      height: auto;
    ",
            "toString": [Function],
          }
        }
        onClick={[Function]}
        src="https://ssl-static.libsyn.com/p/assets/a/4/8/f/a48f1a0697e958ce/Cover_2.png"
      />
    </div>
    <div
      css={
        Object {
          "map": undefined,
          "name": "1r27l9a",
          "next": undefined,
          "styles": "
      display: grid;
      grid-template: auto / auto;
      margin: 10px 0 0;
    ",
          "toString": [Function],
        }
      }
    >
      <div
        css={
          Object {
            "map": undefined,
            "name": "opyjht",
            "next": undefined,
            "styles": "
      grid-column: 1 / 1;
    ",
            "toString": [Function],
          }
        }
      >
        <h3
          css={
            Object {
              "map": undefined,
              "name": "1yuwxk0",
              "next": undefined,
              "styles": "
      margin: 0;
      font-weight: 700;
      text-transform: capitalize;
      color: #000;
      font-size: 0.825rem;
    ",
              "toString": [Function],
            }
          }
          onClick={[Function]}
        >
          Test
        </h3>
      </div>
    </div>
  </div>
</Memo(VideoPreview)>
`;
